File: app/src/main/java/fr/nihilus/xenobladechronicles/datastore/MonsterDao.java
package fr.nihilus.xenobladechronicles.datastore;

import android.arch.lifecycle.LiveData;
import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Delete;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Update;

import fr.nihilus.xenobladechronicles.model.Monster;

@Dao
public interface MonsterDao {

    @Query("SELECT * FROM monster ORDER BY level")
    LiveData<Monster[]> getAll();

    @Query("SELECT * FROM monster WHERE name LIKE :searchedName")
    LiveData<Monster[]> searchByName(String searchedName);

    @Query("SELECT DISTINCT kind FROM monster WHERE kind LIKE :search")
    LiveData<String[]> searchKinds(String search);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(Monster... monster);

    @Update
    void update(Monster monster);

    @Delete
    void delete(Monster monster);
}

File: app/src/main/java/fr/nihilus/xenobladechronicles/datastore/MonsterDao.kt
package fr.nihilus.xenobladechronicles.datastore

import androidx.room.*
import fr.nihilus.xenobladechronicles.model.Monster
import kotlinx.coroutines.flow.Flow

@Dao
interface MonsterDao {

    @Query("SELECT * FROM monster ORDER BY level")
    fun getAll(): Flow<List<Monster>>

    @Query("SELECT * FROM monster WHERE name LIKE :searchedName")
    fun searchByName(searchedName: String): Flow<List<Monster>>

    @Query("SELECT DISTINCT kind FROM monster WHERE kind LIKE :search")
    fun searchKinds(search: String): Flow<List<String>>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(monster: Monster)

    @Update
    suspend fun update(monster: Monster)

    @Delete
    suspend fun delete(monster: Monster)
}
